File: zombie-survivor/typescript/src/Game.ts
import type { Clock } from './Clock.js';
import type { HistoryEntry } from './HistoryEntry.js';
import { DuplicateSurvivorNameException } from './Exceptions.js';
import { Level, maxLevel } from './Level.js';
import { Skill } from './Skill.js';
import { Survivor } from './Survivor.js';

export class Game {
  private readonly _clock: Clock;
  private readonly _survivors: Survivor[] = [];
  private readonly _history: HistoryEntry[] = [];
  private _gameLevel: Level = Level.Blue;
  private _ended = false;

  constructor(clock: Clock) {
    this._clock = clock;
    this.recordEvent('Game started.');
  }

  get survivors(): readonly Survivor[] { return this._survivors; }
  get history(): readonly HistoryEntry[] { return this._history; }
  get gameLevel(): Level { return this._gameLevel; }
  get hasEnded(): boolean { return this._ended; }

  addSurvivor(survivor: Survivor): void {
    if (this._survivors.some(s => s.name === survivor.name)) {
      throw new DuplicateSurvivorNameException(
        `A survivor named '${survivor.name}' already exists.`,
      );
    }
    this._survivors.push(survivor);
    this.recordEvent(`Survivor added: ${survivor.name}.`);
  }

  woundSurvivor(survivor: Survivor): void {
    const wasAlive = survivor.isAlive;
    const changed = survivor.receiveWound();
    if (!changed) return;

    this.recordEvent(`Wound received: ${survivor.name}.`);

    if (wasAlive && !survivor.isAlive) {
      this.recordEvent(`Survivor died: ${survivor.name}.`);
      this.checkGameEnd();
    }
  }

  equipSurvivor(survivor: Survivor, itemName: string): void {
    survivor.equip(itemName);
    this.recordEvent(`Equipment acquired: ${survivor.name} picked up ${itemName}.`);
  }

  killZombie(survivor: Survivor): void {
    const previousXp = survivor.experience;
    survivor.killZombie();
    const levelUp = survivor.checkLevelUp(previousXp);
    if (levelUp !== null) {
      this.recordEvent(`Level up: ${survivor.name} reached ${levelUp}.`);
      if (levelUp === Level.Yellow) {
        survivor.unlockSkill(Skill.PlusOneAction);
      }
      this.updateGameLevel();
    }
  }

  private updateGameLevel(): void {
    let highest = Level.Blue;
    for (const s of this._survivors) {
      if (s.isAlive) highest = maxLevel(highest, s.level);
    }
    if (highest !== this._gameLevel) {
      this._gameLevel = highest;
      this.recordEvent(`Game level changed to ${this._gameLevel}.`);
    }
  }

  private checkGameEnd(): void {
    if (this._survivors.length > 0 && this._survivors.every(s => !s.isAlive)) {
      this._ended = true;
      this.recordEvent('Game ended.');
    }
  }

  private recordEvent(description: string): void {
    this._history.push({ timestamp: this._clock.now(), description });
  }
}

File: zombie-survivor/python/src/zombie_survivor/game.py
from .clock import Clock
from .exceptions import DuplicateSurvivorNameException
from .history_entry import HistoryEntry
from .level import Level, max_level
from .skill import Skill
from .survivor import Survivor


class Game:
    def __init__(self, clock: Clock) -> None:
        self._clock = clock
        self._survivors: list[Survivor] = []
        self._history: list[HistoryEntry] = []
        self._game_level = Level.BLUE
        self._ended = False
        self._record_event("Game started.")

    @property
    def survivors(self) -> list[Survivor]:
        return list(self._survivors)

    @property
    def history(self) -> list[HistoryEntry]:
        return list(self._history)

    @property
    def game_level(self) -> Level:
        return self._game_level

    @property
    def has_ended(self) -> bool:
        return self._ended

    def add_survivor(self, survivor: Survivor) -> None:
        if any(s.name == survivor.name for s in self._survivors):
            raise DuplicateSurvivorNameException(
                f"A survivor named '{survivor.name}' already exists."
            )
        self._survivors.append(survivor)
        self._record_event(f"Survivor added: {survivor.name}.")

    def wound_survivor(self, survivor: Survivor) -> None:
        was_alive = survivor.is_alive
        changed = survivor.receive_wound()
        if not changed:
            return
        self._record_event(f"Wound received: {survivor.name}.")
        if was_alive and not survivor.is_alive:
            self._record_event(f"Survivor died: {survivor.name}.")
            self._check_game_end()

    def equip_survivor(self, survivor: Survivor, item_name: str) -> None:
        survivor.equip(item_name)
        self._record_event(f"Equipment acquired: {survivor.name} picked up {item_name}.")

    def kill_zombie(self, survivor: Survivor) -> None:
        previous_xp = survivor.experience
        survivor.kill_zombie()
        level_up = survivor.check_level_up(previous_xp)
        if level_up is not None:
            self._record_event(f"Level up: {survivor.name} reached {level_up.value}.")
            if level_up == Level.YELLOW:
                survivor.unlock_skill(Skill.PLUS_ONE_ACTION)
            self._update_game_level()

    def _update_game_level(self) -> None:
        highest = Level.BLUE
        for s in self._survivors:
            if s.is_alive:
                highest = max_level(highest, s.level)
        if highest != self._game_level:
            self._game_level = highest
            self._record_event(f"Game level changed to {self._game_level.value}.")

    def _check_game_end(self) -> None:
        if self._survivors and all(not s.is_alive for s in self._survivors):
            self._ended = True
            self._record_event("Game ended.")

    def _record_event(self, description: str) -> None:
        self._history.append(HistoryEntry(self._clock.now(), description))
